proposals/partial-type-inference.md

```diff
@@ -0,0 +1,743 @@
+
+## Motivation
+[motivation]: #motivation
+
+* The current method type inference has two possible areas for improvement.
+  The first improvement regards the strength of the method type inference, which uses only arguments' types to deduce the method's type arguments.
+  Concretely, we can see the weakness in cases where type arguments only depend on target type or type parameter restrictions.
+
+  > Example
+  >
+  > ```csharp
+  > object data = database.fetch();
+  > int count = data.Field("count"); // Error, method type inference can't infer the return type. We have to specify the type argument.
+  > 
+  > public static class Extensions
+  > {
+  >     public static TReturn Field<TReturn>(this object inst, string fieldName) { ... }
+  > }
+  > ```
+  >
+  > ```csharp
+  > test(new MyData()); // Error, method type inference can't infer T. We have to specify all type arguments.
+  >
+  > public void test<T, U>(U data) where T : TestCaseDefault<U> { ... }
+  > ```
+
+  The second improvement regards the "all or nothing" principle, where the method type inference infers either all of the type arguments or nothing.
+
+  > Example
+  > ```csharp
+  > log<, object>(new Message( ... ), null); // Error, impossible to specify just one type argument. We have to specify all of them. 
+  >
+  > public void log<T, U>(T message, U appendix) { ... }
+  > ```
+
+  The first improvement, which would improve the method type inference algorithm, has a significant disadvantage of the breaking change.
+  On the other hand, the second improvement, which would enable specifying some of the method's type arguments, does not influence old code, solves problems regarding the "all or nothing" principle, and reduces the first weakness.
+  
+  > Example
+  > 
+  > ```csharp
+  > test<TestCaseDefault<MyData>, _>(new MyData()); // We can use _ to mark type arguments which should be inferred by the compiler.
+  >
+  > public void test<T, U>(U data) where T : TestCaseDefault<U> { ... }
+  > ```
+  >
+  > ```csharp
+  > log<_, object>(new Message( ... ), null); // We can use _ to mark type arguments which should be inferred by the compiler.
+  >
+  > public void log<T, U>(T message, U appendix) { ... }
+  > ```
+
+* The next motivation is constructor type inference. 
+  Method type inference is not defined on *object_creation_expression*, prohibiting taking advantage of type inference.
+  We divide use cases into the following categories, where type inference would help the programmer. 
+
+  1. Cases where the method type inference would success.
+   
+  > Example
+  > 
+  > ```csharp
+  > var wrappedData = Create(new MyData());
+  > 
+  > public static Wrapper<T> Create<T>(T item) { return new Wrapper<T>(item); }
+  > 
+  > class Wrapper<T> { public Wrapper(T item) { ... } }
+  > ```
+  
+  2. Cases where the method type inference would be weak. (Using type info from target type, or type arguments' constrains)
```

Comment:
```suggestion
  2. Cases where the method type inference would be weak. (Using type info from target type, or type constraints)
```

typo: constrains -> constraints, also applies in other locations below

Reply:
Should be fixed by commit [33b71f0](https://github.com/dotnet/csharplang/pull/7582/commits/33b71f0c489eaa03018d18a9cabf8b2cdf74a43a)